experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+//! 3. `Box` requires an allocation, which is a luxury not all implementors may have, when
+//!    `#![no_std]`  is used.
+//!
+//! Even if we disregard (3), then (2) will get us.  The following snippet compiles, but
+//! forever fixes the associated iteration type to `String`.  Any implementors that don't
+//! implement variants as strings are out of luck.
+//!
+//! ```rust
+//! use core::iter::ExactSizeIterator;
+//! pub trait Trait {
+//!     // Oops... associated type is fixed now.
+//!     fn variants() -> Box<dyn ExactSizeIterator<Item=String>>;
+//! }
+//! ```
+//!
+//! Trying to be clever with genericizing the type also gets us nowhere.  This works, but
+//! requires a `Box`, which in turn requires a change to the type signature of
+//! `LanguageIdentifier::variants()`.
+//!
+//! ```rust
+//! use core::iter::ExactSizeIterator;
+//! pub trait Trait {
+//!     type IterItem;
+//!     fn variants() -> Box<dyn ExactSizeIterator<Item=Self::IterItem>>;
+//! }
+//! ```
+//!
+//! Getting to a generic iterator that doesn't require a box is a piece of gymnastics.  Getting
+//! to an iterable of owned strings requires the iterating trait to be implemented over
+//! *a lifetime-scoped reference* of the implementing type.
+//!
+//! ```rust
+//! pub trait Variants {
+//!     /// The type of the item yieled by the iterator returned by [Variants::variants].  Note
+//!     /// that [Variants::Variant] may be a reference to the type stored in the iterator.
+//!     type Variant;
+//!     /// The type of the iterator returned by [Variants::variants].
+//!     type Iter: ExactSizeIterator<Item = Self::Variant>;
+//!     fn variants(self) -> Self::Iter;
+//! }
+//!
+//! // Here's how to implement the trait when the underlying type is borrowed.
+//!
+//! pub struct BorrowedVariant {
+//!     variants: Vec<&'static str>,
+//! }
+//!
+//! impl<'a> Variants for &'a BorrowedVariant {
+//!     type Variant = &'a &'a str;
+//!     type Iter = std::slice::Iter<'a, &'a str>;
+//!     fn variants(self) -> Self::Iter {
+//!       self.variants.iter()
+//!     }
+//! }
+//!
+//! let borrowed = BorrowedVariant{ variants: vec!["a", "b"], };
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     borrowed.variants()
+//!         .map(|v| v.to_owned().to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
+//! // Here is how to implement the trait when the underlying type is owned.
+//!
+//! pub struct OwnedVariant {
+//!     variants: Vec<String>,
+//! }
+//!
+//! impl<'a> Variants for &'a OwnedVariant {
+//!     type Variant = &'a String;
+//!     type Iter = std::slice::Iter<'a, String>;
+//!     fn variants(self) -> Self::Iter {
+//!         self.variants.iter()
+//!     }
+//! }
+//!
+//! let owned = OwnedVariant{ variants: vec!["a".to_string(), "b".to_string()], };
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     owned.variants()
+//!         .map(|v| v.to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
```

Comment:
Yeah, that's tricky because in Rust, pardon my oversimplification, you either provide an iterator over references, or a collection of values.

See my generic comment for more details.

Reply:
I think this has been handled.